### PR TITLE
Detailed port-in-use error

### PR DESCRIPTION
**Pull request by alexb5dh**
## Changes

Provide details regarding port configuration when binding fails due to port being in use. Example:
```
08 Aug 09:07:24 | Init failed Nethermind.Network.PortInUseException: Port 5432 is in use. If you intend to run 2 or more nodes on one machine, ensure you have changed all configured ports under: 
        Network.P2PPort
        Network.DiscoveryPort
        JsonRpc.WebSocketsPort
        JsonRpc.Port
        JsonRpc.EnginePort
 ---> System.Net.Sockets.SocketException (10048): Only one usage of each socket address (protocol/network address/port) is normally permitted.
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, Boolean disconnectOnFailure, String callerName)
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
   at DotNetty.Transport.Channels.Sockets.TcpServerSocketChannel.DoBind(EndPoint localAddress)
   at DotNetty.Transport.Channels.AbstractChannel.AbstractUnsafe.BindAsync(EndPoint localAddress)
--- End of stack trace from previous location ---
   at DotNetty.Transport.Bootstrapping.AbstractBootstrap`2.DoBindAsync(EndPoint localAddress)
   at Nethermind.Network.NetworkHelper.HandlePortTakenError[T](Func`1 action, Int32[] ports)
   --- End of inner exception stack trace ---
   at Nethermind.Network.NetworkHelper.HandlePortTakenError[T](Func`1 action, Int32[] ports)
```

#### Requires testing
- [x] No

#### If yes, did you write tests?
- [x] No

#### Requires explanation in Release Notes
- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Network/NetworkHelper.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            exception = exception.InnerException!;
+
+        switch (exception)
+        {
+            case SocketException { SocketErrorCode: SocketError.AddressAlreadyInUse or SocketError.AccessDenied }:
+                return ports != null ? new(exception, ports) : new(exception, urls!);
+            case IOException { Source: "Grpc.Core" } when exception.Message.Contains("Failed to bind port"):
+                return ports != null ? new(exception, ports) : new(exception, urls!);
+            default:
+                ExceptionDispatchInfo.Throw(exception);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/NetworkHelper.cs:26]: Nice!

**Review by LukaszRozmej — COMMENTED** [review]
Would be good to indicate which port conflict resulted in this error.

File: src/Nethermind/Nethermind.Network/NetworkHelper.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        {
+            return action();
+        }
+        catch (Exception exception)
+        {
+            throw MapOrRethrow(exception, ports: ports);
+        }
+    }
+
+    public static async Task HandlePortTakenError(Func<Task> action, params string[] urls)
```
[the remark this answers is not in the document]
Comment (alexb5dh) [reply src/Nethermind/Nethermind.Network/NetworkHelper.cs:55]: Tried initially using extensions, but it required around 5-6 different signature, due to each binding using a different method and didn't see it making code any simpler. Can return to it though.

[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/NetworkHelper.cs:55]: If you think it didn't work, then ignore it. Maybe you can post here how invocation would look like?

**Comment by alexb5dh** [comment]
> Would be good to indicate which port conflict resulted in this error.
@LukaszRozmej, it already does, but by port number as you can see in the PR description.

I decided to not include specific option name due to some of the calls not having a single port to blame. For example, `WebHost.Start` in `JsonRpcRunner` can involve 2 or 3 ports depending on the configuration and there's no way to extract exact conflicting one.

Regarding showing all the options. Idea was that since the issue is likely to happen due to an attempt to run another instance on the same machine, it would be better to show all the configurable ports, instead of showing a new (failed one) each run.

**Review by LukaszRozmej — APPROVED** [review]

**Comment by LukaszRozmej** [comment]
@alexb5dh can you resolve conflicts?

**Comment by alexb5dh** [comment]
@LukaszRozmej, done

**Pull request merged**